### PR TITLE
Query Refactoring: moving validation to setters and constructors

### PR DESCRIPTION
**Pull request by cbuescher**
Moving validation from validate() to constructors and setters for the following query builders:
- GeoDistanceQueryBuilder
- GeoDistanceRangeQueryBuilder
- GeoPolygonQueryBuilder
- GeoShapeQueryBuilder
- GeohashCellQuery
- TermsQueryBuilder

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java [hunk 1]
```diff
@@ -76,12 +75,7 @@
     /** Whether or not to accept geo points with invalid latitude or longitude */
     private boolean ignoreMalformed = DEFAULT_IGNORE_MALFORMED;
 
-    static final GeoDistanceQueryBuilder PROTOTYPE = new GeoDistanceQueryBuilder();
-
-    /** For serialization purposes only. */
-    private GeoDistanceQueryBuilder() {
-        this.fieldName = null;
-    }
+    static final GeoDistanceQueryBuilder PROTOTYPE = new GeoDistanceQueryBuilder("field");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java]: I used the empty-string or a single whitespace instead

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java [hunk 2]
```diff
@@ -191,6 +178,9 @@ public GeoDistanceQueryBuilder optimizeBbox(String optimizeBbox) {
         if (Strings.isEmpty(optimizeBbox)) {
             throw new IllegalArgumentException("optimizeBbox must not be null or empty");
         }
+        if (false == (optimizeBbox.equals("none") || optimizeBbox.equals("memory") || optimizeBbox.equals("indexed"))) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java]: you can use switch statements here instead ie.
```Java
switch(optimzeBox) { 
    case "none":
    case "memory":
    case "indexed":
      break;
    default: 
         throw new IllegalArgumentException("optimizeBbox must be one of [none, memory, indexed]");
}
```

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java [hunk 3]
```diff
@@ -92,6 +98,13 @@ public GeoPoint point() {
     }
 
     public GeoDistanceRangeQueryBuilder from(Object from) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java]: why dont' we have `from(String s)` `from(Number)` and let the compiler decide

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java [hunk 4]
```diff
@@ -101,6 +114,13 @@ public Object from() {
     }
 
     public GeoDistanceRangeQueryBuilder to(Object to) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java]: see my comment above?

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java [hunk 5]
```diff
@@ -175,6 +190,12 @@ public DistanceUnit unit() {
     }
 
     public GeoDistanceRangeQueryBuilder optimizeBbox(String optimizeBbox) {
+        if (optimizeBbox == null) {
+            throw new IllegalArgumentException("optimizeBox must not be null");
+        }
+        if (false == (optimizeBbox.equals("none") || optimizeBbox.equals("memory") || optimizeBbox.equals("indexed"))) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java]: see what I recommended in the ctor and impl it here. Then you can just all the setter from the ctor

File: core/src/main/java/org/elasticsearch/indices/cache/query/terms/TermsLookup.java [hunk 6]
```diff
@@ -96,6 +89,37 @@ public TermsLookup routing(String routing) {
...
+        String id = null;
+        String path = null;
+        String routing = null;
+        XContentParser.Token token;
+        String currentFieldName = null;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if ("index".equals(currentFieldName)) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/cache/query/terms/TermsLookup.java]: maybe use a switch / case here?

**Comment by s1monw** [comment]
looks good, I left some comments

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java]: We don't allow for empty String in most constructors (at least fieldname), whitespace would work but decided against it to make it clearer what the arguments are. Can change this of course if there's other reasons.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryBuilder.java]: maybe just use `_na_`

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java]: The only reason I can think of is that also lt(), lte(), gt() and gte() would also need two versions, quiet a lot of setters then. We can also remove those, rely on the API user to explicitely call includeUpper/Lower then, removes convenience though.

**Comment by cbuescher** [comment]
Adressed the comments, squashed and rebased.

File: core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryParser.java [hunk 7]
```diff
@@ -205,20 +202,20 @@ public GeoDistanceRangeQueryBuilder fromXContent(QueryParseContext parseContext)
...
+            if (vFrom instanceof Number) {
+                queryBuilder.from((Number) vFrom);
+            } else {
+                queryBuilder.from((String) vFrom);
+            }
         }
 
         if (vTo != null) {
-            queryBuilder.to(vTo);
+            if (vTo instanceof Number) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryParser.java]: you can do this simpler and just call `vTo.toString()` this should work just fine?

**Comment by s1monw** [comment]
left a minor comment LGTM - not need for another review

**Pull request merged**